Build guide: validation receipts (#553)
holochain/docs-pages: src/pages/build/validation-receipts.md

---
title: "Validation Receipts"
---

::: intro

> An agent can get a sense of the DHT availability of their data by asking their conductor for the **validation receipts** it's collected. These receipts are created by the peers that the agent has sent their [**DHT operations**](/build/dht-operations/) to as an acknowledgement of valid data.

Reviewer: ```suggestion
An agent can get a sense of the DHT availability of their data by asking their conductor for the **validation receipts** it's collected. These receipts are created by the peers that the agent has sent their [**DHT operations**](/build/dht-operations/) to as a confirmation that they have validated the data.
```

> An agent can get a sense of the DHT availability of their data by asking their conductor for the **validation receipts** it's collected. These receipts are created by the peers that the agent has sent their [**DHT operations**](/build/dht-operations/) to as an acknowledgement of valid data.

Reviewer: > An agent can get a sense of the DHT availability

More accurate to say something like "an agent can get a sense of how many others they have published their data to", and then to clarify that these receipts are only collected in response to publishing, so it excludes any peers who received the data via gossip

> An agent can get a sense of the DHT availability of their data by asking their conductor for the **validation receipts** it's collected. These receipts are created by the peers that the agent has sent their [**DHT operations**](/build/dht-operations/) to as a confirmation that they have validated the data.

Reviewer: ```suggestion
An agent can get a sense of the DHT availability of their data by asking their conductor for the **validation receipts** it's collected. These receipts are created by the peers that the agent has published their [**DHT operations**](/build/dht-operations/) to as a confirmation that they have validated the data.
```

Author: https://github.com/holochain/docs-pages/pull/553/commits/1ceeacb9901b3a2d36497b1151cbdbc718d3478c

:::

> As described in the [DHT operations](/build/dht-operations/) page, each action that an agent authors is turned into a set of DHT operations that are sent to other agents around the network for validation. If an operation is found to be valid, it'll transform the state of the DHT at the operation's [**basis address**](/resources/glossary/#basis-address). At this point, the validator will also send back a validation receipt to the author.

Reviewer: I thought you only send validation receipts in response to publish? So not when you receive them via gossip.

Author: Yes, that's something I mention elsewhere (at least i think I do); would it be better to mention it here instead?

Reviewer: Yes I just want to make super explicit that the validation receipts are only coming from peers you published to. This makes it sound like anyone who receives the Op will send a validation receipt to the author.

Author: Did some work on this https://github.com/holochain/docs-pages/pull/553/commits/a83e985b3854d7fe6b57eab4ea07c30682a140e9

> These validation receipts helps the author's conductor keep track of how fully their data has 'saturated' into the DHT --- that is, how many other agents know about it and can serve it up to anyone who asks for it. The purpose is to help the conductor decide whether it needs to try publishing it to more peers --- it'll keep trying until it collects the number of receipts specified in the [`required_validations` argument](/build/entries/#required-validations) passed to the `entry_type` macro (default is 5).

Reviewer: ```suggestion
These validation receipts helps the author's conductor keep track of how fully their data has been published to the DHT --- that is, how many other agents they have published to and can serve it up to anyone who asks for it. The purpose is to help the conductor decide whether it needs to try publishing it to more peers --- it'll keep trying until it collects the number of receipts specified in the [`required_validations` argument](/build/entries/#required-validations) passed to the `entry_type` macro (default is 5).
```

## Get validation receipts

> To get validation receipts for an action that an agent has authored, use the [`get_validation_receipts`](https://docs.rs/hdk/latest/hdk/validation_receipt/fn.get_validation_receipts.html) host function. It takes an input consisting of the action hash and returns a result containing a vector of [`ValidationReceiptSet`](https://docs.rs/hdk/latest/hdk/prelude/struct.ValidationReceiptSet.html) values. Each of these items corresponds to one of the operations produced from the action and gives details on that operation's DHT status from the perspective of the author.

Reviewer: ```suggestion
To get validation receipts for an action that an agent has authored, use the [`get_validation_receipts`](https://docs.rs/hdk/latest/hdk/validation_receipt/fn.get_validation_receipts.html) host function. It takes an input consisting of the action hash and returns a result containing a vector of [`ValidationReceiptSet`](https://docs.rs/hdk/latest/hdk/prelude/struct.ValidationReceiptSet.html) values. Each of these items corresponds to one of the operations produced from the action and gives details on that operation's DHT status from the perspective of the validating agent.
```

Author: Hm, I think here it would be the perspective of the author, because it's the set of receipts they've collected -- the perspective of the validator is seen in each of the items in the `ValidationReceiptSet::receipts` vec. But it seems that's not clear from the language I used -- is this better? https://github.com/holochain/docs-pages/pull/553/commits/7c3ba938ef24f2fd0e24c0a246eb73de1c684c79

    let validation_receipt_set = get_validation_receipts(GetValidationReceiptsInput { action_hash })?;

    let worst_status = validation_receipt_set
        .iter()
        .flat_map(
            |set| set.receipts.iter()
                .map(|receipt| receipt.validation_status)
        )
        .fold(
            ValidationStatus::Valid,
            |acc, status| match status {
                ValidationStatus::Valid => acc,
                // Validation can be abandoned if dependencies can't be fetched.
                ValidationStatus::Abandoned =>
                    // Rejected is worse than abandoned.
                    // But it also should never happen if the agent is self-
                    // validating their DHT ops before publishing them (except
                    // in the case of an invalid membrane proof -- see
                    // https://developer.holochain.org/build/genesis-self-check-callback/#the-need-for-basic-pre-validation )
                    if acc == ValidationStatus::Rejected { acc }
                    else { status },
                ValidationStatus::Rejected => status,
            }
        );

    Ok(worst_status)
}
```

> This example imagines a 'saturation score' for authored data; this could be used by the front end to warn a user that their peers might not yet be able to see their most recent database contributions. {#saturation-score}

Author: This example might between walking the line between basic and unnecessarily contrived, but as I see it, this is the main use case for `get_validation_receipts`. Willing to take feedback on it -- I just don't want to present a function without demonstrating why people should care about it.

Reviewer: I'm okay with the function, but I don't think "saturation" is the right term. It's really publishing progress.

Author: Ah! Great suggestion; I didn't like 'saturation' but my creative juices weren't pumping enough to come up with something else. Edited in https://github.com/holochain/docs-pages/pull/553/commits/29b2a640220a910f0fc67e8ecac3cf0ef163fa82

    // Now calculate the score, which will be the number of receipts we have
    // divided by the number of receipts we expected to have.
    // First we need to know how many we expect to have.
    // This is the default number of required receipts for a given action.
    // For app entries, the default can be changed and can be gotten with
    // the `dna_info` host function; see
    // https://developer.holochain.org/build/entries/#required_validations
    let full_saturation_receipt_count = 5;
    // The number of ops varies by the action; the receipt set will have one
    // element per op type for the action type, even if there are no receipts
    // collected for a given op type.
    let number_of_ops_for_this_action = validation_receipt_set.len();
    let number_of_expected_receipts = full_saturation_receipt_count * number_of_ops_for_this_action;

    let total_receipts_collected = validation_receipt_set
        .iter()
        .fold(0, |acc, set_for_op| acc + set_for_op.receipts.len());

    Ok(total_receipts_collected as f32 / number_of_expected_receipts as f32)
}
```

> ## Cautions and design guidelines

Reviewer: I think this should all be removed except for the noted line that is specifically relevant to this function.

A DHT is an [**eventually consistent**](https://en.wikipedia.org/wiki/Eventual_consistency) database, which means that not every agent has the exact same up-to-date view of the shared data. This has some important consequences:

* Writing to the DHT isn't a "did or didn't happen" event. A state change lives on a gradient from "it didn't happen" to "everybody has seen and integrated it".
* This state changes over time; agents will keep on trying to publish their authored operations until they receive the number of receipts they expected.
* An author may not receive validation receipts in a timely manner if network conditions are poor, even if the validators they published to have completed their work.

> * Only the conductor hosting the agent who authored an action will get the validation receipts. That means that only the authoring cell, and other cells with the same DNA on the same conductor, will be able to access them with `get_validation_receipts`.

Reviewer: This line is important to note.

* Because people are accustomed to centralized data stores in which everyone sees the same state at roughly the same time, we recommend that you design user experiences that expose the concept of eventual consistency in ways that people can easily understand. The [saturation score example](#saturation-score) above might be a useful start.

## Reference

* [`hdk::validation_receipt::get_validation_receipts`](https://docs.rs/hdk/latest/hdk/validation_receipt/fn.get_validation_receipts.html)
* [`holochain_zome_types::validate::ValidationReceiptSet`](https://docs.rs/holochain_zome_types/latest/holochain_zome_types/validate/struct.ValidationReceiptSet.html)
* [`holochain_zome_types::validate::ValidationReceiptInfo`](https://docs.rs/holochain_zome_types/latest/holochain_zome_types/validate/struct.ValidationReceiptInfo.html)

## Further reading

* [Build Guide: Validation](/build/validation/)
* [Build Guide: DHT Operations](/build/dht-operations/)
* [Core Concepts: DHT](/concepts/4_dht/)
* [Core Concepts: Validation](/concepts/7_validation/)

> * [Wikipedia: Eventual Consistency](https://en.wikipedia.org/wiki/Eventual_consistency)

Reviewer: I don't think this or the subsequent bullet belong here.

Author: how come?

Reviewer: Because its not specifically related to validation receipts. Every building block of holochain is related to eventual consistency, but we don't want to link to it on every page. If we want to talk about it, I'd say it fits when talking about the DHT.

Author: ah well that's true -- useful info but not necessarily germane to this page. Removed in https://github.com/holochain/docs-pages/pull/553/commits/c2e60e6595cfb8ee90ee87e7886b50d289c76a9b

---
title: "Validation Receipts"
---

::: intro
An agent can get a rough sense of the DHT availability of their data by checking how many **validation receipts** it's collected. These receipts are created by the peers that the agent has **published** their [**DHT operations**](/build/dht-operations/) to as a confirmation that they have validated the data and are now serving it.
:::

> As described in the [DHT operations](/build/dht-operations/) page, each action that an agent authors is turned into a set of DHT operations that are published to other agents in the network for validation. If an operation is found to be valid, it'll transform the state of the DHT at the operation's [**basis address**](/resources/glossary/#basis-address). At this point, the validator will also send back a validation receipt to the author.

Author: accidentally fixed in https://github.com/holochain/docs-pages/pull/553/commits/fdf6243612afdb3ab6f2f0452a948ca927ccc1f6#diff-48f02a1ce35d1e03bb6b65a999456dd257a4b96d93e63d02eea644a6ff980b81L9-R9

> These validation receipts helps the author's conductor keep track of how fully their data has been published to the DHT --- that is, how many other agents they have published to and can serve it up to anyone who asks for it. The purpose is to help the conductor decide whether it needs to try publishing it to more peers --- it'll keep trying until it collects the number of receipts specified in the [`required_validations` argument](/build/entries/#required-validations) passed to the `entry_type` macro (default is 5).

Reviewer 2: I think that's okay as a shorthand in most cases. It's necessary to not describe all details at all times. As long as that process is described in detail somewhere.

This example checks to see if any validator has abandoned or rejected an operation for an action:

```rust
use hdk::prelude::*;

fn check_validation_status(action_hash: ActionHash) -> ExternResult<ValidationStatus> {
    let validation_receipt_set = get_validation_receipts(GetValidationReceiptsInput { action_hash })?;

    let worst_status = validation_receipt_set
        .iter()
        .flat_map(
            |set| set.receipts.iter()
                .map(|receipt| receipt.validation_status)
        )
        .fold(
            ValidationStatus::Valid,
            |acc, status| match status {
                ValidationStatus::Valid => acc,
                // Validation can be abandoned if dependencies can't be fetched.

>                 ValidationStatus::Abandoned =>

Author: lovely. The fewer bits of API that I don't understand the purpose of but have to come up with some way to document, the better. Removed in https://github.com/holochain/docs-pages/pull/553/commits/4e321fbcb6268e0ab66b2c780af8efe5ff253c7e

>                 ValidationStatus::Abandoned =>

Author: re-added and fixed in https://github.com/holochain/docs-pages/pull/553/commits/56aa1e5913f70f6be0cc371b0304d95675722a9c

                    // Rejected is worse than abandoned.
                    // But it also should never happen if the agent is self-
                    // validating their DHT ops before publishing them (except
                    // in the case of an invalid membrane proof -- see
                    // https://developer.holochain.org/build/genesis-self-check-callback/#the-need-for-basic-pre-validation )
                    if acc == ValidationStatus::Rejected { acc }
                    else { status },
                ValidationStatus::Rejected => status,
            }
        );

    Ok(worst_status)
}
```

This example gives a simple yes/no answer to whether an action's operations have been published to the DHT in a sufficient amount.

This function could be used by the front end to warn a user that their peers might not yet be able to see their most recent database contributions.

```rust
use hdk::prelude::*;

#[hdk_extern]

> pub fn has_action_been_fully_published(action_hash: ActionHash) -> ExternResult<bool> {

Author: lovely. The fewer bits of API that I don't understand the purpose of but have to come up with some way to document, the better. Removed in https://github.com/holochain/docs-pages/pull/553/commits/4e321fbcb6268e0ab66b2c780af8efe5ff253c7e

By default, every DHT operation produced by an action must collect five validation receipts before the author considers publishing to be complete. For application entry creation actions, you can override this by setting the [`required_validations`](/build/entries/#required_validations) field on the entry type.
!!!

But publishing isn't an all-or-nothing event; it happens over time. Here's an example that gives more nuanced feedback on publishing progress.

<!-- TODO/FIXME: currently this function will give inaccurate results. Fix this if https://github.com/holochain/holochain/issues/4861 gets resolved -->

```rust
use hdk::prelude::*;

#[hdk_extern]
pub fn calculate_publish_progress_score(action_hash: ActionHash) -> ExternResult<f32> {
    let validation_receipt_set = get_validation_receipts(GetValidationReceiptsInput { action_hash })?;

    // First figure out how many receipts we need to collect.

    // The default expectation is five receipts per op. Note that, if this is
    // an app entry creation action and you've set `required_validations` to
    // something other than 5 for the entry type, this will give an
    // inaccurate score.

>     let expected_op_receipt_count = 5;

Reviewer 2: It's interesting that you can't get at the required count in the HDK. I wonder if the macro might generate something undocumented. Either way, a ticket is a good solution

}
```

> This example gives a simple yes/no answer to whether an action's operations have been published to the DHT in a sufficient amount.

Author: @mattyg I'd love a review of this new code -- realised there was no example that showed you could just check the 'complete' count for a really coarse check. Do you think it adds?

pub fn calculate_publish_progress_score(action_hash: ActionHash) -> ExternResult<f32> {
    let validation_receipt_set = get_validation_receipts(GetValidationReceiptsInput { action_hash })?;

>     // Now calculate the score, which will be the number of receipts we have

Author: @mattyg I've also updated the comments here; I think it didn't flow very well before.
